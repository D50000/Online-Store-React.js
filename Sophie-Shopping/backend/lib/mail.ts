import { createTransport, getTestMessageUrl } from "nodemailer";

const transporter = createTransport({
  host: "smtp.ethereal.email",
  port: 587,
  auth: {
    user: "donny.beatty34@ethereal.email",
    pass: "shxZASXrt7NC2paQTv",
  },
});

function createSimpleEmail(text: string): string {
  return `
    <div className="email" style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>Hi, Sophie Shopping's User !!! </p>
    </div>
  `;
}

export interface MailResponse {
  accepted?: string[] | null;
  rejected?: null[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: Envelope;
  messageId: string;
}

export interface Envelope {
  from: string;
  to?: string[] | null;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transporter.sendMail({
    to,
    from: "test@test.com",
    subject: "Your password reset token!",
    html: createSimpleEmail(`Your Password Reset Token Is Here!
    <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click Here to reset</a>
  `),
  })) as MailResponse;

  if (process.env.MAIL_USER.includes("ethereal.email")) {
    console.log(`ðŸ’Œ Message Sent!  Preview it at ${getTestMessageUrl(info)}`);
  }
}
