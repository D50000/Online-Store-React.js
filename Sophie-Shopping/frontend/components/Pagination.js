import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import Head from 'next/head';
import Link from 'next/link';
import PropTypes from 'prop-types';

import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    _allProductsMeta {
      count
    }
  }
`;

export default function Pagination({ page }) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY, {
    fetchPolicy: 'no-cache',
  });
  if (loading) return 'Loading ........';
  if (error) return <DisplayError error={error} />;

  console.log(data._allProductsMeta);
  const { count } = data._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles>
      <Head>
        <title>
          Page {page} of {pageCount}
        </title>
      </Head>
      <Link href={`/products/${page - 1}`}>
        {/* can't pull property aria-disabled in <Link> */}
        <a aria-disabled={page <= 1}>-- Previous</a>
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`}>
        <a aria-disabled={page >= pageCount}>Next ++</a>
      </Link>
    </PaginationStyles>
  );
}

Pagination.propTypes = {
  page: PropTypes.number,
};
