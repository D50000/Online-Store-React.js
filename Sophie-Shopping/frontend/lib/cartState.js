import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // Custom provider can access it via the consumer.

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// Make the custom hook for these functions.
function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };

CartStateProvider.propTypes = {
  children: PropTypes.any,
};
