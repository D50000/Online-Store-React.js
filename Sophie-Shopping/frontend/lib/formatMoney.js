export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'TWD',
    minimumFractionDigits: 0,
  };

  const formatter = Intl.NumberFormat('zh-US', options);

  return formatter.format(amount);
}
