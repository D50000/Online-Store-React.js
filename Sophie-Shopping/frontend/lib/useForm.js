import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial);

  function handleChange(e) {
    // Handle the data type that pass in.
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      value[0] = e.target.files;
    }
    setInputs({
      ...inputs,
      // Dynamic update the key and value
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const emptyState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );
    setInputs(emptyState);
  }

  // Return the initialState and SetStateAction()
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
