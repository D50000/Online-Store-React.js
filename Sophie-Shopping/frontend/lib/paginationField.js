import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // Setup Apollo keys myself.
    read(existing = [], { args, cache }) {
      // console.log({ existing, args, cache });
      const { skip, first } = args;

      // Read the detail of items on the page from the cache.
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // Check if we have existing items.
      const items = existing.slice(skip, skip + first).filter((item) => item);

      // On the last page.
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      // No item, fetch it from the DB.
      if (items.length !== first) {
        return false;
      }
      if (items.length) {
        return items;
      }

      // Fallback to network
      return false;
    },
    // Apollo fetch back the data will run merge()
    merge(existing, incoming, { args }) {
      const { skip, first } = args;

      // Calculate the item need to update.
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; i + 1) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
