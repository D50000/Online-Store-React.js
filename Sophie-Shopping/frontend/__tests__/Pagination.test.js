import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';

import Pagination from '../components/Pagination';
import { makePaginationMocksFor } from '../lib/testUtils';

describe('<Pagination />', () => {
  it('display a loading message', () => {
    const { container } = render(
      <MockedProvider mocks={makePaginationMocksFor(1)}>
        <Pagination />
      </MockedProvider>
    );
    expect(container).toHaveTextContent('Loading');
  });

  // Single page is 4 items.
  it('render pagination for 18 items', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    expect(container).toHaveTextContent('Page 1 of 5');

    const pageCountSpan = screen.getByTestId('pageCount');
    expect(pageCountSpan).toHaveTextContent('5');
    expect(container).toMatchSnapshot();
  });

  it('disable the prev page on first page', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(10)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    const preButton = screen.getByText(/Previous/); // Regular Expression
    const nextButton = screen.getByText(/Next/);

    expect(preButton).toHaveAttribute('aria-disabled', 'true');
    expect(nextButton).toHaveAttribute('aria-disabled', 'false');
    expect(container).toMatchSnapshot();
  });

  it('disable the next page on last page', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(10)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    const preButton = screen.getByText(/Previous/); // Regular Expression
    const nextButton = screen.getByText(/Next/);

    expect(preButton).toHaveAttribute('aria-disabled', 'false');
    expect(nextButton).toHaveAttribute('aria-disabled', 'true');
    expect(container).toMatchSnapshot();
  });

  it('enables both on middle page', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(10)}>
        <Pagination page={2} />
      </MockedProvider>
    );
    await screen.findByTestId('pagination');
    const preButton = screen.getByText(/Previous/); // Regular Expression
    const nextButton = screen.getByText(/Next/);

    expect(preButton).toHaveAttribute('aria-disabled', 'false');
    expect(nextButton).toHaveAttribute('aria-disabled', 'false');
    expect(container).toMatchSnapshot();
  });
});
