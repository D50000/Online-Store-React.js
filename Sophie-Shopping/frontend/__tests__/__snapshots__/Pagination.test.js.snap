// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> disable the next page on last page 1`] = `
<div>
  <div
    class="PaginationStyles-sc-zbbm55-0 jzlRep"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/2"
    >
      -- Previous
    </a>
    <p>
      Page 
      3
       of 
      <span
        data-testid="pageCount"
      >
        3
      </span>
    </p>
    <p>
       
      10
       Items Total
    </p>
    <a
      aria-disabled="true"
      href="/products/4"
    >
      Next ++
    </a>
  </div>
</div>
`;

exports[`<Pagination /> disable the prev page on first page 1`] = `
<div>
  <div
    class="PaginationStyles-sc-zbbm55-0 jzlRep"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/products/0"
    >
      -- Previous
    </a>
    <p>
      Page 
      1
       of 
      <span
        data-testid="pageCount"
      >
        3
      </span>
    </p>
    <p>
       
      10
       Items Total
    </p>
    <a
      aria-disabled="false"
      href="/products/2"
    >
      Next ++
    </a>
  </div>
</div>
`;

exports[`<Pagination /> enables both on middle page 1`] = `
<div>
  <div
    class="PaginationStyles-sc-zbbm55-0 jzlRep"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/1"
    >
      -- Previous
    </a>
    <p>
      Page 
      2
       of 
      <span
        data-testid="pageCount"
      >
        3
      </span>
    </p>
    <p>
       
      10
       Items Total
    </p>
    <a
      aria-disabled="false"
      href="/products/3"
    >
      Next ++
    </a>
  </div>
</div>
`;

exports[`<Pagination /> render pagination for 18 items 1`] = `
<div>
  <div
    class="PaginationStyles-sc-zbbm55-0 jzlRep"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/products/0"
    >
      -- Previous
    </a>
    <p>
      Page 
      1
       of 
      <span
        data-testid="pageCount"
      >
        5
      </span>
    </p>
    <p>
       
      18
       Items Total
    </p>
    <a
      aria-disabled="false"
      href="/products/2"
    >
      Next ++
    </a>
  </div>
</div>
`;
